### PR TITLE
Sync native calls

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/IISIntegration/issues/1006

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs [hunk 1]
```diff
@@ -182,7 +182,7 @@ CancellationToken IHttpRequestLifetimeFeature.RequestAborted
 
         bool IHttpResponseFeature.HasStarted => HasResponseStarted;
 
-        bool IHttpUpgradeFeature.IsUpgradableRequest => _server.IsWebSocketAvailable(_pInProcessHandler);
+        bool IHttpUpgradeFeature.IsUpgradableRequest => true;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs]: Don't we still need this check?

File: src/Microsoft.AspNetCore.Server.IIS/Core/IO/AsyncIOEngine.cs [hunk 2]
```diff
@@ -23,8 +24,9 @@ internal partial class AsyncIOEngine : IAsyncIOEngine
         private AsyncWriteOperation _cachedAsyncWriteOperation;
         private AsyncFlushOperation _cachedAsyncFlushOperation;
 
-        public AsyncIOEngine(IntPtr handler)
+        public AsyncIOEngine(object contextSync, IntPtr handler)
         {
+            this._contextSync = contextSync;
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IO/AsyncIOEngine.cs]: nit drop the this.

File: src/Microsoft.AspNetCore.Server.IIS/Core/IO/WebSocketsAsyncIOEngine.cs [hunk 3]
```diff
@@ -100,7 +112,10 @@ private void ThrowIfNotInitialized()
 
         public void Dispose()
         {
-            NativeMethods.HttpTryCancelIO(_handler);
+            lock (_contextLock)
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IO/WebSocketsAsyncIOEngine.cs]: Why do you need to lock here?

**Comment by jkotalik** [comment]
For #1006

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs (hunk 1 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs]: We reset feature to null if websockets are disabled (https://github.com/aspnet/IISIntegration/blob/3bdb33ccd4dbe5baa3162ca3d3778df1883db093/src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.cs#L151) so if user was able to get the feature request is upgradeable.

File: src/Microsoft.AspNetCore.Server.IIS/Core/IO/WebSocketsAsyncIOEngine.cs (hunk 3 above)
Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IO/WebSocketsAsyncIOEngine.cs]: Locking all IO

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IO/WebSocketsAsyncIOEngine.cs]: We don't know that it's safe to call HttpTryCancelIO overlapped with other calls on the HttpContext

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs [hunk 4]
```diff
@@ -205,7 +205,10 @@ ClaimsPrincipal IHttpAuthenticationFeature.User
                     return null;
                 }
 
-                return NativeMethods.HttpTryGetServerVariable(_pInProcessHandler, variableName, out var value) ? value : null;
+                lock (_contextLock)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs]: Add a comment explaining why we take the lock here.

**Comment by pakrym** [comment]
> functional tests

I can add a test that tries to force the crash but it would be inherently racy and unreliable so I'm a bit sceptical about it.

**Comment by jkotalik** [comment]
As I mentioned offline, at worst you can verify that the behavior that normally would be racy doesn't fail. Ex send a request to the server that obtains server variables and verify it doesn't crash.

**Comment by pakrym** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.cs [hunk 5]
```diff
@@ -241,7 +240,7 @@ private void EnsureIOInitialized()
             // If at this point request was not upgraded just start a normal IO engine
             if (AsyncIO == null)
             {
-                AsyncIO = new AsyncIOEngine(_pInProcessHandler);
+                AsyncIO = new AsyncIOEngine(_contextLock, _pInProcessHandler);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.cs]: Seems bad to have a shared lock between the HttpContext and AsyncIOEngine. Worth it to move OnStarting and OnCompleted to the AsyncIOEngine?

Comment (pakrym) [reply src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.cs]: `OnStarting` and `OnCompleted` don't really belong to IOEngine having multiple lock objects might have been more correct but it's just more allocations.

**Pull request merged**